api/control/v1alpha1/types.go

```diff
@@ -302,6 +315,211 @@ type BindingList struct {
+}
+
+// NamedAggregator pairs a name with a way to aggregate over some objects.
+//
+// - For `type=="COUNT"`, `subject` is omitted and the aggregate is the count
+// of those objects that are not `null`.
+//
+// - For the other types, `subject` is required and SHOULD
+// evaluate to a numeric value; exceptions are handled as follows.
+// For a string value: if it parses as an int64 or float64 then that is used.
+// Otherwise this is an error condition: a value of 0 is used, and the error
+// is reported in the BindingPolicyStatus.Errors (not necessarily repeated for each WEC).
+type NamedAggregator struct {
+	Type AggregatorType `json:"type"`
+
+	// +optional
+	Subject *Expression `json:"subject,omitempty"`
+}
+
+// AggregatorType indicates what sort of aggregation is to be done.
+type AggregatorType string
+
+const (
+	AggregatorTypeCount AggregatorType = "COUNT"
+	AggregatorTypeSum   AggregatorType = "SUM"
+	AggregatorTypeAvg   AggregatorType = "AVG"
+	AggregatorTypeMin   AggregatorType = "MIN"
+	AggregatorTypeMax   AggregatorType = "MAX"
+)
+
+// Expression is some value to derive from a workload object from a WEC.
+// An Expression is either a JSONPath identifying a value to extract
+// or a boolean combination of comparisons of atomic values.
+// While the expression here in the Go type system admits other values,
+// the controller will accept only the restricted set stated above.
+type Expression struct {
+	Op   ExpressionOperator `json:"op"`
+	Path string             `json:"path,omitempty"`
+	Args []Expression       `json:"args,omitempty"`
+}
+
+type ExpressionOperator string
+
+const (
+	OperatorPath  ExpressionOperator = "Path" // JSONPath string
+	OperatorOr    ExpressionOperator = "Or"
+	OperatorAnd   ExpressionOperator = "And"
+	OperatorNot   ExpressionOperator = "Not"
+	OperatorEqual ExpressionOperator = "Equal"
+)
+
+// StatusCollectorStatus defines the observed state of StatusCollector.
+type StatusCollectorStatus struct {
+	ObservedGeneration int64 `json:"observedGeneration"`
+
+	// +optional
+	Errors []string `json:"errors,omitempty"`
+}
+
+// StatusCollectorList is the API type for a list of StatusCollector.
+//
+// +kubebuilder:object:root=true
+type StatusCollectorList struct {
+	metav1.TypeMeta `json:",inline"`
+	// Standard list metadata.
+	// More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#types-kinds
+	// +optional
+	metav1.ListMeta `json:"metadata,omitempty"`
+	Items           []StatusCollector `json:"items"`
+}
+
+// CombinedStatus holds the combined status from the WECs for one particular (workload object, BindingPolicy) pair.
+// The namespace of the CombinedStatus object is the namespace of the workload object,
+// or "kubestellar-report" if the workload object has no namespace.
+// The name of the CombinedStatus object is the concatenation of:
+// - the NamespacedName the workload object
+// - the string ":"
+// - the Name of the BindingPolicy object.
```

Comment:
(A) NamespacedName is not a string, the reader wonders how it is concatenated into a string.
(B) For any reasonable answer to (A), the prescribed combination might overflow the limit on name length.